src/transformers/models/seamless_m4t/modeling_seamless_m4t.py

```diff
@@ -0,0 +1,4058 @@
+            )
+        return reordered_past
+
+
+@add_start_docstrings(
+    "The text-to-speech SeamlessM4T Model transformer which can be used for T2ST.",
+    SEAMLESS_M4T_START_DOCSTRING,
+)
+class SeamlessM4TForTextToSpeech(SeamlessM4TPreTrainedModel):
+    _keys_to_ignore_on_load_missing = ["final_logits_bias", "speech_encoder"]
+
+    def __init__(self, config):
+        super().__init__(config)
+
+        self.input_model = SeamlessM4TMultiModalToTextModelWithLMHead(
+            config, use_text_encoder=True, use_speech_encoder=False
+        )
+
+        self.t2u_model = SeamlessM4TTextToUnitWithLMHead(config)
+
+        # Initialize weights and apply final processing
+        self.post_init()
+
+    def get_encoder(self):
+        return self.input_model.model.text_encoder
+
+    def get_decoder(self):
+        return self.input_model.get_decoder()
+
+    def get_input_embeddings(self):
+        return self.input_model.get_input_embeddings()
+
+    # @add_start_docstrings_to_model_forward(SEAMLESS_M4T_INPUTS_DOCSTRING.format("batch_size, sequence_length"))
+    # @add_code_sample_docstrings(
+    #    checkpoint=_CHECKPOINT_FOR_DOC,
+    #    output_type=BaseModelOutputWithPastAndCrossAttentions,
+    #    config_class=_CONFIG_FOR_DOC,
+    # )
+    def forward(
+        self,
+        input_ids: torch.LongTensor = None,
+        attention_mask: Optional[torch.Tensor] = None,
+        decoder_input_ids: Optional[torch.LongTensor] = None,
+        decoder_attention_mask: Optional[torch.LongTensor] = None,
+        head_mask: Optional[torch.Tensor] = None,
+        decoder_head_mask: Optional[torch.Tensor] = None,
+        cross_attn_head_mask: Optional[torch.Tensor] = None,
+        encoder_outputs: Optional[Tuple[Tuple[torch.FloatTensor]]] = None,
+        past_key_values: Optional[Tuple[Tuple[torch.FloatTensor]]] = None,
+        inputs_embeds: Optional[torch.FloatTensor] = None,
+        decoder_inputs_embeds: Optional[torch.FloatTensor] = None,
+        labels: Optional[torch.LongTensor] = None,
+        use_cache: Optional[bool] = None,
+        output_attentions: Optional[bool] = None,
+        output_hidden_states: Optional[bool] = None,
+        return_dict: Optional[bool] = None,
+        **kwargs,
+    ) -> Union[Seq2SeqLMOutput, Tuple[torch.FloatTensor]]:
+        r"""
+        labels (`torch.LongTensor` of shape `(batch_size, sequence_length)`, *optional*):
+            Labels for computing the masked language modeling loss. Indices should either be in `[0, ...,
+            config.vocab_size]` or -100 (see `input_ids` docstring). Tokens with indices set to `-100` are ignored
+            (masked), the loss is only computed for the tokens with labels in `[0, ..., config.vocab_size]`.
+
+        Returns:
+
+        """
+
+        logger.warning(
+            "This calls `self.input_model.forward`. If you want to generate speech, use the `generate` method."
```

Comment:
So this is the text-to-speech model comprised of **two** encoder-decoder models, concatenated together:
1. Text encoder-decoder
2. Text-to-units encoder-decoder

Am I right in saying that this model is **not** trained end-to-end? i.e. we train the text encoder-decoder separately to the text-to-unit encoder-decoder?

Note that if it **is** trained end-to-end, we need a notion of a forward pass, such that we can run it end-to-end in a training script

Otherwise, if it's inference only, it's going to be quite hard to define a sensible forward pass. I think then we have two options:
1. Only do the forward pass for the text encoder-decoder
2. Skip the forward pass entirely 

=> I'd be in favour of 1 (as you have here). Keen to hear your thoughts @amyeroberts @ArthurZucker

Reply:
> Am I right in saying that this model is not trained end-to-end?

Yes exactly (if I'm not mistaken) !
And I agree for going to 1., unlike Bark for example, I wanted to be able to properly test the models that generate speech (the T2S, the S2S and the SeamlessM4TModel as well).

However, most of the `GenerationTesterMixin` tests won't pass anyways because they rely on comparing `.generate` (which call this override generate) with the different generation strategies (e.g `.greedy_search` which rely on the `forward`). So my reasons for going with 1. might no longer hold! WDYT?